lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/EvaluatorRuntime.cs

```diff
@@ -231,17 +235,31 @@ private void OnException(Exception e)
                 Logger.Log(Level.Error, "Evaluator {0} failed with exception {1}.", _evaluatorId, e);
                 _state = State.FAILED;
 
-                var errorMessage = string.Format(
-                        CultureInfo.InvariantCulture,
-                        "failed with error [{0}] with message [{1}] and stack trace [{2}]",
-                        e,
-                        e.Message,
-                        e.StackTrace);
+                byte[] errorBytes = null;
+
+                try
+                {
+                    using (var memStream = new MemoryStream())
+                    {
+                        var formatter = new BinaryFormatter();
+                        formatter.Serialize(memStream, e);
+                        errorBytes = memStream.ToArray();
+                    }
+                }
+                catch (SerializationException se)
+                {
+                    using (var memStream = new MemoryStream())
+                    {
+                        var formatter = new BinaryFormatter();
+                        formatter.Serialize(memStream, new NonSerializableEvaluatorException(e.ToString(), se));
```

Comment:
The idea is to have the intended use of the class captured in the factory method, and then have the constructors themselves `private`. That way, we can change the constructors in the future. I don't think it is terribly important here, hence we can skip this.